pkgs/development/compilers/cudatoolkit/redist/find_outputs/find_outputs/outputs.py

```diff
@@ -0,0 +1,311 @@
+import logging
+import multiprocessing
+from itertools import chain
+from pathlib import Path
+from typing import (
+    Iterator,
+    Literal,
+    Sequence,
+    TypeAlias,
+    get_args,
+)
+
+from pydantic import BaseModel, Field
+
+from find_outputs import generic, manifest
+from find_outputs.utilities import (
+    NixStoreEntry,
+    is_nonempty,
+    nix_store_prefetch_file,
+    nix_store_unpack_archive,
+)
+
+Output: TypeAlias = Literal[
+    "bin",
+    "dev",
+    "doc",
+    "lib",
+    "static",
+    "sample",
+]
+
+
+class ReleaseFeatures(BaseModel):
+    root_dirs: Sequence[str] = Field(alias="rootDirs")
+    has_bin: bool = Field(False, alias="hasBin")
+    has_dev: bool = Field(False, alias="hasDev")
+    has_doc: bool = Field(False, alias="hasDoc")
+    has_lib: bool = Field(False, alias="hasLib")
+    has_static: bool = Field(False, alias="hasStatic")
+    has_sample: bool = Field(False, alias="hasSample")
+
+    def get_outputs(self) -> list[Output]:
+        return [
+            output
+            for output in get_args(Output)
+            if getattr(self, f"has_{output}", False)
+        ]
+
+
+Package: TypeAlias = generic.Package[ReleaseFeatures]
+
+
+def get_dynamic_libraries(path: Path) -> Iterator[Path]:
+    """
+    Returns a list of dynamic libraries in the directory tree of the given path.
+    """
+    return (
+        entry
+        for entry in chain(
+            *(path.rglob("*" + suffix) for suffix in {".so", ".so.*", ".dylib", ".dll"})
+        )
+        if entry.is_file()
+    )
```

Comment:
a number of functions in this file seem like they are ripe for abstraction with a function like
```py
def file_paths_matching(regexes):
  return lambda path: (
    entry
    for entry in chain(*(path.rglob(regex) for regex in regexes))
    if entry.is_file()
  )

# then, eg,
get_dynamic_libraries = file_paths_matching({"*.so", "*.so.*", "*.dylib", "*.dll"})
```

* `get_dynamic_libraries`
* `get_static_libraries`
* `get_executables` (with an extra filter)
* `get_cmake_modules`
* `get_headers`
* `get_python_modules`

can then all become one-liners

Reply:
See https://github.com/ConnorBaker/cuda-redist-find-features/commit/371e11a0e10335a3ffd17e109353b40a3903d150